contrib/postgresql16/template.py

```diff
@@ -0,0 +1,198 @@
+_major = "16"
+_default_ver = True  # should this version provide non-versioned resources?
+pkgname = f"postgresql{_major}"
+pkgver = f"{_major}.2"
+pkgrel = 0
+# NOTE: version 16 doesn't work with meson + tarball
+# switch to meson for version 17
+build_style = "gnu_configure"
+_bindir = f"usr/libexec/{pkgname}"
+_datadir = f"usr/share/{pkgname}"
+_includedir = "usr/include/postgresql"
+_srvlibdir = f"usr/lib/{pkgname}"
+configure_args = [
+    f"--bindir=/{_bindir}",
+    f"--includedir=/{_includedir}",
+    f"--datadir=/{_datadir}",
+    "--disable-rpath",
+    # "--with-llvm", # NOTE: postgresql 16 doesn't support llvm 16+
+    "--with-libxml",
+    "--with-ssl=openssl",
+    "--with-uuid=e2fs",
+    "--with-perl",
+    "--with-python",
+    "--with-tcl",
+    "--with-lz4",
+    "--with-zstd",
+    "--with-system-tzdata=/usr/share/zoneinfo",
+]
+configure_gen = []
+make_cmd = "gmake"
+make_build_target = "world"
+hostmakedepends = [
+    "gmake",
+    "pkgconf",
+]
+makedepends = [
+    "e2fsprogs-devel",
+    "icu-devel",
+    "libxml2-devel",
+    "linux-headers",
+    "lz4-devel",
+    "openssl-devel",
+    "perl",
+    "python-devel",
+    "readline-devel",
+    "tcl-devel",
+    "zlib-devel",
+    "zstd-devel",
+]
+depends = ["postgresql-common", "tzdata"]
+provides = ["postgresql"]
+pkgdesc = "Sophisticated object-relational DBMS"
+maintainer = "mia <mia@mia.jetzt>"
+license = "PostgreSQL"
+url = "https://www.postgresql.org"
+source = [
+    f"https://ftp.postgresql.org/pub/source/v{pkgver}/postgresql-{pkgver}.tar.bz2"
+]
+sha256 = ["446e88294dbc2c9085ab4b7061a646fa604b4bec03521d5ea671c2e5ad9b2952"]
+# checks depend on libpq already being installed
+options = ["!check"]
+
+
+def post_install(self):
+    if _default_ver:
+        self.install_man("doc/src/sgml/man*/*", glob=True)
+    self.install_file(self.files_path / "pltcl_create_tables.sql", "usr/share")
+    (self.destdir / _bindir / "PG_VERSION").write_text(_major)
+
+
+@subpackage(f"{pkgname}-devel")
+def _devel(self):
+    return [f"usr/lib/{pkgname}/pgxs"]
+
+
+def _contrib_search(self, check):
+    return filter(
+        check,
+        map(lambda p: p.name, (self.parent.cwd / "build/contrib").iterdir()),
+    )
+
+
+def _subpackage_install(self, path):
+    self.parent.do(
+        "gmake",
+        "-C",
+        f"build/{path}",
+        f"DESTDIR={self.chroot_destdir}",
+        "install",
+    )
+
+
+@subpackage(f"{pkgname}-contrib")
+def _contrib(self):
+    def subpackage():
+        _subpackage_install(self, "contrib")
+        for subdir in _contrib_search(
+            self,
+            lambda name: name.endswith("_plperl") or name.endswith("_plpython"),
+        ):
+            self.parent.do(
```

Comment:
running commands in subpackage install hooks violates packaging guidelines, subpackage install hooks should *only* steal files

Reply:
an exception we have is for libreoffice that does the install file stuff. i think here it's "almost" similar in that all of these have a large amount of files (that change per version seemingly) that are laborious to manually separate instead of calling gmake install per subpackage. is there a way that doesn't involve listing a million files?

Reply:
fwiw file lists can also be used anywhere else if necessary, the important part is that they are generated ahead of time